### PR TITLE
[REEF-320]:Change reflection to tang instantiation in Wake Layer for Writables

### PR DESCRIPTION
**Pull request by dkm2110**
[REEF-320]:Change reflection to tang instantiation in Wake Layer for Writables

This addressed the issue by
  * Taking Iinjector as argument in RemoteManager and Transport Layer
  * Instantiating messages by tang injection than by reflection

JIRA: [REEF-320](https://issues.apache.org/jira/browse/REEF-320)

### CODE REVIEW

**Comment by dkm2110** [comment]
@markusweimer @jwang98052 @markusweimer 

I have introduced Iinjectors in Wake Layer. This is a straightforward change. Please review and merge if all is well.

File: lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableLink.cs [hunk 1]
```diff
@@ -174,7 +174,8 @@ public T Read()
                 return default(T);
             }
 
-            T value = _cache.GetInstance(dataType);
+            T value = (T) _injector.ForkInjector().GetInstance(dataType);
+            //T value = _cache.GetInstance(dataType);
 
             if (value == null)
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableLink.cs]: This should never happen. You can either throw in this case or wait for the null exception to happen.

You could, however, catch `InjectionException` from the `GetInstance` call and use `default(T)` in those cases.

File: lang/cs/Org.Apache.REEF.Wake.Tests/PrefixedWritableString.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    {
+    }
+    
+    
+    /// <summary>
+    /// Writable wrapper around the string class which takes integer prefix
+    /// This class is used to test non empty injector in TransportServer and Client
+    /// </summary>
+    [Obsolete("Need to remove Iwritable and use IstreamingCodec. Please see Jira REEF-295 ", false)]
+    public class PrefixedWritableString : IWritable
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Wake.Tests/PrefixedWritableString.cs]: Nitpick: Rename to `PrefixedStringWritable`

File: lang/cs/Org.Apache.REEF.Wake.Tests/PrefixedWritableString.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+namespace Org.Apache.REEF.Wake.Tests
+{
+    [NamedParameter("identifier in PrefixedWritable")]
+    public class StringId : Name<int>
+    {
+    }
+    
+    
+    /// <summary>
+    /// Writable wrapper around the string class which takes integer prefix
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Wake.Tests/PrefixedWritableString.cs]: The int isn't used in serializing, right?

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Wake.Tests/PrefixedWritableString.cs]: Yes, it is not used.

File: lang/cs/Org.Apache.REEF.Wake.Tests/PrefixedWritableString.cs (hunk 2 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Wake.Tests/PrefixedWritableString.cs]: done

File: lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableLink.cs (hunk 1 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableLink.cs]: done

**Comment by dkm2110** [comment]
@markusweimer 

I have pushed new code after changes suggested by you.

Dhruv

File: lang/cs/Org.Apache.REEF.Wake.Tests/PrefixedStringWritable.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Wake.Remote;
+
+namespace Org.Apache.REEF.Wake.Tests
+{
+    [NamedParameter("identifier in PrefixedWritable")]
+    public class StringId : Name<int>
+    {
+    }
+
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Wake.Tests/PrefixedStringWritable.cs]: remove empty line

File: lang/cs/Org.Apache.REEF.Wake.Tests/WritableRemoteManagerTest.cs [hunk 5]
```diff
@@ -156,6 +167,58 @@ public void TestWritableTwoWayCommunication()
...
 
         /// <summary>
+        /// Tests two way communications where message needs an injectable argument 
+        /// to be passed. Checks whether both sides are able to receive messages
+        /// </summary>
+        [TestMethod]
+        public void TestNonEmptyArgumentInjectionWritableTwoWayCommunication()
+        {
+            IPAddress listeningAddress = IPAddress.Parse("127.0.0.1");
+
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Wake.Tests/WritableRemoteManagerTest.cs]: remove empty line

File: lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableLink.cs [hunk 6]
```diff
@@ -174,15 +174,16 @@ public T Read()
                 return default(T);
             }
 
-            T value = _cache.GetInstance(dataType);
-
-            if (value == null)
+            try
+            {
+                T value = (T) _injector.ForkInjector().GetInstance(dataType);
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableLink.cs]: Why fork here?

Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableLink.cs]: How do you guarantee the dataType is a interface or a class?

File: lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableRemoteEvent.cs [hunk 7]
```diff
@@ -72,7 +78,7 @@ public WritableRemoteEvent()
         /// <param name="reader">The reader from which to read </param>
         public void Read(IDataReader reader)
         {
-            Value = Activator.CreateInstance<T>();
+            Value = (T) _injector.ForkInjector().GetInstance(typeof(T));
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableRemoteEvent.cs]: remove space after <T>

Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableRemoteEvent.cs]: Are you expecting different instance every time this method is called even T is the same?

File: lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableTransportClient.cs [hunk 8]
```diff
@@ -44,12 +46,14 @@ public class WritableTransportClient<T> : IDisposable where T : IWritable
...
-        public WritableTransportClient(IPEndPoint remoteEndpoint)
+        /// <param name="injector">The injector to pass arguments to incoming messages</param>
+        public WritableTransportClient(IPEndPoint remoteEndpoint, IInjector injector)
         {
             Exceptions.ThrowIfArgumentNull(remoteEndpoint, "remoteEndpoint", Logger);
 
-            _link = new WritableLink<T>(remoteEndpoint);
+            _link = new WritableLink<T>(remoteEndpoint, injector);
             _cancellationSource = new CancellationTokenSource();
+            _injector = injector;
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableTransportClient.cs]: Is the _injector used anywhere else that you need to keep an instance variable?

File: lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableTransportServer.cs [hunk 9]
```diff
@@ -212,29 +219,26 @@ public void Dispose(bool disposing)
...
                         break;
                     }
 
                     TransportEvent<T> transportEvent = new TransportEvent<T>(message, link);
-
                     _remoteObserver.OnNext(transportEvent);
                 }
                 LOGGER.Log(Level.Error,
                     "ProcessClient close the Link. IsCancellationRequested: " + token.IsCancellationRequested);
+
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableTransportServer.cs]: remove

File: lang/cs/Org.Apache.REEF.Wake.Tests/PrefixedStringWritable.cs (hunk 4 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Wake.Tests/PrefixedStringWritable.cs]: done

File: lang/cs/Org.Apache.REEF.Wake.Tests/WritableRemoteManagerTest.cs (hunk 5 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Wake.Tests/WritableRemoteManagerTest.cs]: done

File: lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableLink.cs (hunk 6 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableLink.cs]: You always want to get new instantiation. Thats why.

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableLink.cs]: .GetType() always give class type an not interface type

File: lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableRemoteEvent.cs (hunk 7 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableRemoteEvent.cs]: Yes we need different instance.

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableRemoteEvent.cs]: done

File: lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableTransportClient.cs (hunk 8 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableTransportClient.cs]: No, I have removed it.

File: lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableTransportServer.cs (hunk 9 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableTransportServer.cs]: done

**Comment by dkm2110** [comment]
@jwang98052  @markusweimer 
I have incorporated julias and markus ocmments. If everything is fine please merge.

**Comment by markusweimer** [comment]
I'll test and merge

**Pull request closed without merging**